src/integer_mod_q/mat_zq/solve.rs

```diff
@@ -58,44 +62,314 @@ impl MatZq {
         let mut out = MatZq::new(self.get_num_columns(), 1, &matrix.get_mod());
         for (i, j) in indices.iter() {
-            let entry: Z = matrix.get_entry(*i, matrix.get_num_columns() - 1).ok()?;
-            out.set_entry(*j, 0, &entry).ok()?;
+            let entry: Z = matrix
+                .get_entry(*i, matrix.get_num_columns() - 1)
+                .ok()
+                .unwrap();
+            out.set_entry(*j, 0, &entry).ok().unwrap();
+        }
+
+        Some(out)
+    }
+
+    /// Computes a solution for a system of linear equations under a modulus
+    /// of the form `z^a`.
+    /// It solves `Ax = y` for `x` with `A` being a [MatZq] value.
+    /// If no solution is found, `None` is returned.
+    ///
+    /// Note that this function does not compute a solution whenever there is one.
+    /// If the matrix has not full rank under a modulus that divides the given one,
+    /// `None` may be returned even if the system may be solvable.
+    /// If the matrix has dimensions larger than 50 or the number of columns
+    /// exceeds the number of rows by a factor of 2, this is very unlikely to happen.
+    ///
+    /// Note that this function is a part of `solve_gaussian_elimination` and
+    /// does not check for the correctness of the given parameters.
+    ///
+    /// Parameters:
+    /// - `y`: The syndrome for which a solution has to be computed.
+    /// - `base`: The base of the modulus.
+    /// - `power`: The power of the modulus.
+    ///
+    /// Returns a solution for the linear system or `None`, if none could be computed.
+    fn solve_hensel(&self, y: &MatZq, base: &Z, power: &u64) -> Option<MatZq> {
+        // Set `matrix_base` to the given matrix under `base` as the modulus to
+        // compute a solution for the system under `base` as modulus.
+        let matrix_base = MatZq::from((&MatZ::from(self), &Modulus::from(base)));
+
+        // Concatenate the matrix with the identity matrix under `base`
+        // as its modulus to apply gaussian elimination on it.
+        let mut matrix_identity_base_gauss = matrix_base
+            .concat_horizontal(&MatZq::identity(
+                self.get_num_rows(),
+                self.get_num_rows(),
+                base,
+            ))
+            .unwrap();
+
+        // Saves the indices of row and column, where we created a 1 entry
+        // such that we do not have to go through the matrix afterwards.
+        let mut indices: Vec<(i64, i64)> = Vec::new();
+        let mut row_count = 0;
+
+        // Saves the permutation of the gaussian elimination.
+        let mut permutation: Vec<i64> = vec![];
+        for i in 0..self.get_num_rows() {
+            permutation.push(i);
+        }
+        for column_nr in 0..self.get_num_columns() {
+            if !indices.is_empty() && indices[indices.len() - 1].0 >= self.get_num_rows() {
+                break;
+            }
+            let used_rows: Vec<i64> = indices.iter().map(|(row, _)| *row).collect();
```

Comment:
maybe update used_rows and indices simultaneously such that you do not have to compute used_rows for each column again